src/core/imports/mpi_collectives.hpp

```diff
@@ -0,0 +1,407 @@
+/*
+   Copyright (c) 2009-2016, Jack Poulson
+   All rights reserved.
+
+   Copyright (c) 2013, Jeff Hammond
+   All rights reserved.
+
+   Copyright (c) 2013, Jed Brown
+   All rights reserved.
+
+   This file is part of Elemental and is under the BSD 2-Clause License,
+   which can be found in the LICENSE file in the root directory, or at
+   http://opensource.org/licenses/BSD-2-Clause
+*/
+#include <El-lite.hpp>
+#include "mpi_utils.hpp"
+#ifdef HYDROGEN_USES_ALUMINUM
+#include "Al.hpp"
+#ifdef HYDROGEN_USES_NCCL2
+#include "nccl_impl.hpp"
+#endif
+#endif //HYDROGEN_USES_ALUMINUM
+
+typedef unsigned char* UCP;
+
+namespace El {
+namespace mpi {
+
+#ifdef HYDROGEN_USES_ALUMINUM
+#ifndef HYDROGEN_USES_NCCL2
+using backend = ::Al::MPIBackend;
+#else
+using backend = ::Al::NCCLBackend;
+#endif
+
+template<typename Real>
+void aluminum_allreduce( const Real* sbuf, Real* rbuf, int count, Op op, Comm comm )
+{
+    MPI_Op mpi_op = op.op;
+    Al::ReductionOperator red_op = Al::internal::mpi::MPI_Op2ReductionOperator (mpi_op);
+
+    if(TypeMap<Real>() == MPI_UNSIGNED_CHAR){
```

Comment:
Same comment as below. This whole `if...else if...else` should just be a single line:

```Al::Allreduce<backend>(sbuf, rbuf, count, red_op, *comm.aluminum_comm)```

Reply:
This was only way I knew how to bypass the error (see below) caused by something like ValueInt, Entry, and Complex, which I reported in a previous meeting (probably around PR#7). Please advise me if there is an easier way to deal with this compile error.

 [ 29%] Linking CXX shared library libHydrogen.so
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<long>(long const*, long*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<long>()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<unsigned long>(unsigned long const*, unsigned long*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<unsigned long>()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<El::ValueInt<int> >(El::ValueInt<int> const*, El::ValueInt<int>*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<El::ValueInt<int> >()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<El::Entry<int> >(El::Entry<int> const*, El::Entry<int>*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<El::Entry<int> >()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<El::ValueInt<float> >(El::ValueInt<float> const*, El::ValueInt<float>*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<El::ValueInt<float> >()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<El::ValueInt<El::Complex<float> > >(El::ValueInt<El::Complex<float> > const*, El::ValueInt<El::Complex<float> >*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<El::ValueInt<El::Complex<float> > >()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<El::Entry<float> >(El::Entry<float> const*, El::Entry<float>*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<El::Entry<float> >()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<El::Entry<El::Complex<float> > >(El::Entry<El::Complex<float> > const*, El::Entry<El::Complex<float> >*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<El::Entry<El::Complex<float> > >()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<El::ValueInt<double> >(El::ValueInt<double> const*, El::ValueInt<double>*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<El::ValueInt<double> >()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<El::ValueInt<El::Complex<double> > >(El::ValueInt<El::Complex<double> > const*, El::ValueInt<El::Complex<double> >*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<El::ValueInt<El::Complex<double> > >()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<El::Entry<double> >(El::Entry<double> const*, El::Entry<double>*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<El::Entry<double> >()'
CMakeFiles/Hydrogen.dir/src/core/imports/mpi.cpp.o: In function `void Al::NCCLBackend::NonblockingAllreduce<El::Entry<El::Complex<double> > >(El::Entry<El::Complex<double> > const*, El::Entry<El::Complex<double> >*, unsigned long, Al::ReductionOperator, Al::NCCLCommunicator&, CUstream_st*&, Al::NCCLCollectiveAlgorithm)':
/p/lscratchf/ayoo/al/Aluminum/install/include/aluminum/nccl_impl.hpp:198: undefined reference to `ncclDataType_t Al::internal::nccl::TypeMap<El::Entry<El::Complex<double> > >()'
collect2: error: ld returned 1 exit status
make[2]: *** [libHydrogen.so] Error 1
make[1]: *** [CMakeFiles/Hydrogen.dir/all] Error 2
make: *** [all] Error 2

Reply:
I'm pretty sure this is because most of these types don't exist in NCCL. Just don't instantiate the template for these types. Actually, don't instantiate the template at all -- client code compile time will go up slightly, but who cares? (Not me.)

Reply:
Actually, the "correct" way to deal with this is probably to have a set of acceptable types (i.e., the "Aluminum types") and only enable this template if `Real` is an acceptable type. After that, the compiler will handle dispatch appropriately. I can do this if you don't want to take on the metaprogramming this will entail. Let me know.